feature: administration api guide (#86)
mojaloop/mojaloop-specification: admin-api/admin-api-specification-v1.0.md

---
showToc: true
---
# Introduction

This document provides detailed information about the Central Ledger API. The Central Ledger API is a Mojaloop API enabling Hub Operators to manage admin processes around:

- creating/activating/deactivating participants in the Hub
- adding and updating participant endpoint information
- managing participant accounts, limits, and positions
- creating Hub accounts
- performing Funds In and Funds Out operations
- creating/updating/viewing settlement models
- retrieving transfer details

 For background information about the participant and settlement model details that the Hub Operator can administer using the Central Ledger API, see section [Basic concepts](#basic-concepts).

## Basic concepts

To provide context for the admin operations that the Central Ledger API enables, this section gives a brief definition of some basic concepts.

**Participant**

> Either the Hub itself or a Digital Financial Service Provider (DFSP) that is a participant in a Mojaloop scheme.

Reviewer: nit: This section is a little hard to read with the formatting - maybe you can add make the heading and content on a different line? Or perhaps use `### Participant` instead of `**Participant**`

Author: OK, will make the change.

**Endpoint**
The DFSP callback URL where the Hub routes API callbacks. The URL specified is the endpoint set up in the outbound API gateway.

**Limit**

> Also called "Net Debit Cap (NDC)". The NDC is a limit or a cap placed on a DFSP's funds available for transacting. It represents the maximum amount of money that the DFSP can owe to other DFSPs. The Position is continuously checked against the Net Debit Cap and when the Position amount exceeds the NDC amount, transfers are blocked.

Reviewer 2: I would clarify that only the "NDC" is supported at this time. The idea of the Limits is to provide extensibility for future limit requirements.

Reviewer 3: I might suggest a precision....

"...if a Transfer would cause the Position amount to exceed the NDC amount, the transfer is blocked."

I believe the formula, (TransferAmount + Position) <= NDC is checked dynamically, it doesn't cause a "bit to flip" and all future transfers to be blocked.

Author: OK, thank you, I have made the change.

Author: Perhaps it's easier to spot where the change is here: https://github.com/mojaloop/mojaloop-specification/pull/86/commits/ab5dfaea1627cf74b0115f3550b2b382dce71398

|**401**|Unauthorized|The request requires authentication in order to be processed.|
|**403**|Forbidden|The request was denied and will be denied in the future.|
|**404**|Not Found|The requested resource is not available at the moment.|
|**405**|Method Not Allowed|An unsupported HTTP method for the request was used.|
|**406**|Not Acceptable|The requested resource is capable of generating only content not acceptable according to the Accept headers sent in the request.|
|**500**|Internal Server Error|A generic error message, given when an unexpected condition was encountered and no more specific message is suitable.|
|**501**|Not Implemented|The server does not support the requested service. The client should not retry.|
|**503**|Service Unavailable|The server is currently unavailable to accept any new service requests. This should be a temporary state, and the client should retry within a reasonable time frame.|

# API services

This section introduces and details all services that the API supports for each resource and HTTP method.

## High-level API services

On a high level, the API can be used to perform the following actions:

> - View, create, update participant-related details, such as limit (Net Debit Cap), position, or endpoints configured.
>   - Use the services provided by the API resource **`/participants`**.
> - View, create, update settlement-model-related details, such as granularity, delay, liquidity check, and so on.
>   - Use the services provided by the API resource **`/settlementModels`**.
> - View transaction details for a particular transfer.
>   - Use the services provided by the API resource **`/transactions`**.

Reviewer: nit: formatting improvements... less boilerplate language
```suggestion
-  **`/participants`**: View, create, update participant-related details, such as limit (Net Debit Cap), position, or endpoints configured.
- **`/settlementModels`**: View, create, update settlement-model-related details, such as granularity, delay, liquidity check, and so on.
- **`/transactions`*: View transaction details for a particular transfer.

```

Author: Yes, agree.

### Supported API services

[Table "Supported API services"](#table-supported-api-services) includes high-level descriptions of the services that the API provides. For more detailed information, see the sections that follow.

> | URI  |  HTTP method `GET`| HTTP method `PUT` | HTTP method `POST` | HTTP method `DELETE` |

Reviewer: You've already said that `DELETE` is not supported above... let's remove it from this table and make it easier to read.

Author: I would like to leave this as is if it's OK with you. Readers don't necessarily read in a linear fashion. Users often scan-read for bits of information they are interested in. A reader who decided to give their attention to this table may not have read the section titled "HTTP methods" (which is where we first stated that `DELETE` is not supported).

Reviewer: Yep, 100% agree.

| `/participants/{name}/accounts/{id}/transfers/{transferId}`  | Not supported  | Not supported  | Record a Transfer as a Funds In or Out transaction for a participant account  |  Not supported |
| `/participants/{name}/initialPositionAndLimits`  | Not supported  | Not supported  |  Add initial participant limits and position | Not supported  |
| **`/settlementModels`**  | View all settlement models  | Not supported  | Create a settlement model  |  Not supported |
| `/settlementModels/{name}`  | View settlement model by name  | Update a settlement model (activate/deactivate a settlement model) | Not supported  |  Not supported |
| **`/transactions/{id}`**  | Retrieve transaction details by `transferId`  | Not supported  | Not supported  | Not supported |

## API Resource `/participants`

The services provided by the resource `/participants` are primarily used by the Hub Operator for viewing, creating, and updating participant-related details, such as limit (Net Debit Cap), position, or endpoints configured.

### `GET /participants`

The `GET /participants` request is used to retrieve information about all participants.

#### Data model

> The `GET /participants` request sends an empty request body.

Reviewer: This section would be infinitely more readable with examples of requests and responses for each endpoint.

Stripe has good examples of this - see https://stripe.com/docs/api/balance/balance_retrieve

Author: Agree, will add those in. My brief said I should follow the conventions employed by the FSPIOP API doc, but I agree, examples definitely add a lot of value.

Reviewer: thanks :)

The following data is returned for each participant in the case of a successful operation:

| Name  | Required  | Type | Description |
|---|---|--|--|
| `name`  |  yes | [String(2..30)](#string) | The name of the participant. |
| `id`  |  yes | [String](#string) | The identifier of the participant in the form of a fully qualified domain name combined with the participant's `fspId`. |
| `created`  |  yes | [DateTime](#datetime)  | Date and time when the participant was created. |
| `isActive`  |  yes | [Integer(1)](#integer) | A flag to indicate whether or not the participant is active. Possible values are `1` and `0`. |
| `links`  |  yes | [Self](#self) | List of links for a Hypermedia-Driven RESTful Web Service. | 	
| `accounts`  |  yes | [Accounts](#accounts) | The list of ledger accounts configured for the participant. |

### `POST /participants`

> The `POST /participants` request is used to create a participant in the Hub.

Reviewer: nit: cut down on redundant language. We are already in the `POST /participants` section, and we already know it's a  HTTP request
```suggestion
Creates a participant in the Hub.

```

Author: OK, agree.

#### Data model

> The `POST /participants` request sends the following data in the request body:
> 
> | Name  | Required  | Type | Description |
> |---|---|--|--|
> | `name`  |  yes | [String(2..30)](#string) | The name of the participant. |
> | `currency`  |  yes | [CurrencyEnum](#currencyenum) | The currency that the participant will transact in. |
> 
> The following data is returned for the newly created participant in the case of a successful operation:
> 
> | Name  | Required  | Type | Description |
> |---|---|--|--|
> | `name`  |  yes | [String(2..30)](#string) | The name of the participant. |
> | `id`  |  yes | [String](#string) | The identifier of the participant in the form of a fully qualified domain name combined with the participant's `fspId`. |
> | `created`  |  yes | [DateTime](#datetime)  | Date and time when the participant was created. |
> | `isActive`  |  yes | [Integer(1)](#integer) | A flag to indicate whether or not the participant is active. Possible values are `1` and `0`. |
> | `links`  |  yes | [Self](#self) | List of links for a Hypermedia-Driven RESTful Web Service. | 	
> | `accounts`  |  yes | [Accounts](#accounts) | The list of ledger accounts configured for the participant. |

Reviewer: Let's cut down on boilerplate language in favour of consistent headings.

```suggestion

##### Parameters

| Name  | Required  | Type | Description |
|---|---|--|--|
| `name`  |  yes | [String(2..30)](#string) | The name of the participant. |
| `currency`  |  yes | [CurrencyEnum](#currencyenum) | The currency that the participant will transact in. |

##### Returns
A newly created participant.

| Name  | Required  | Type | Description |
|---|---|--|--|
| `name`  |  yes | [String(2..30)](#string) | The name of the participant. |
| `id`  |  yes | [String](#string) | The identifier of the participant in the form of a fully qualified domain name combined with the participant's `fspId`. |
| `created`  |  yes | [DateTime](#datetime)  | Date and time when the participant was created. |
| `isActive`  |  yes | [Integer(1)](#integer) | A flag to indicate whether or not the participant is active. Possible values are `1` and `0`. |
| `links`  |  yes | [Self](#self) | List of links for a Hypermedia-Driven RESTful Web Service. | 	
| `accounts`  |  yes | [Accounts](#accounts) | The list of ledger accounts configured for the participant. |

```

... and so on...

Author: I don't agree with the "Parameters" bit. I would prefer to keep the term "parameter" for query parameters or path parameters (that practice is also in sync with Swagger terminology). Having said that, I _will_ cut down on boilerplate language.

Reviewer: Sure. We could have separate table for query or path parameters, or some specs have a `where` column in the table which states where the parameter should be places (path, query, body etc.) 

Mainly I though we could do better than vague terms such as: _"sends the following data "_

This section defines element data types used by the API.

### Amount

For details, see section [7.2.13 Amount](../fspiop-api/documents/API-Definition_v1.1.md#7213-amount) in the Mojaloop FSPIOP API Definition.

### Boolean

A `"true"` or `"false"` value.

### DateTime

For details, see section [7.2.14 DateTime](../fspiop-api/documents/API-Definition_v1.1.md#7214-datetime) in the Mojaloop FSPIOP API Definition.

### Enum

For details, see section [7.2.2 Enum](../fspiop-api/documents/API-Definition_v1.1.md#722-enum) in the Mojaloop FSPIOP API Definition.

### Integer

For details, see section [7.2.5 Integer](../fspiop-api/documents/API-Definition_v1.1.md#725-integer) in the Mojaloop FSPIOP API Definition.

### Number

The API data type **Number** is a an arbitrary-precision, base-10 decimal number value.

### String

For details, see section [7.2.1 String](../fspiop-api/documents/API-Definition_v1.1.md#721-string) in the Mojaloop FSPIOP API Definition.

### UUID

For details, see section [7.2.16 UUID](../fspiop-api/documents/API-Definition_v1.1.md#7216-uuid) in the Mojaloop FSPIOP API Definition.

## Element definitions

This section defines element types used by the API.

> ### AutoPositionReset

Reviewer: Why does this need a separate definition? It's already defined inline in the above data models.

Author: Good point, will remove it.

Data model for the element **AutoPositionReset**.

| Name  | Required  | Type | Description |
|---|---|--|--|
| `autoPositionReset`  | yes | [Boolean](#boolean) | A flag to indicate whether or not a settlement model requires the automatic reset of the position. |

### IsActive

Data model for the element **IsActive**.

| Name  | Required  | Type | Description |
|---|---|--|--|

> | `isActive`  | yes | [Integer(1)](#integer) | A flag to indicate whether or not a ledger account / participant is active. Possible values are `1` (active) and `0` (not active). |

Reviewer: Why do we have `IsActive` and `IsActiveBoolean`? That seems pretty nasty to me.

Author: The API sometimes takes the active status to be indicated by `1`, and at other times by `true`. A distinction between the two had to be made. The body of the `PUT /participants/{name}` sends `"isActive": true`, whereas the response gives you `"isActive": 1`.

Reviewer: Yeah this isn't great API design. If we have a term such as `isActive`, it's data type should be consistent. Why can't we make them consistent?

### IsActiveBoolean

Data model for the element **IsActiveBoolean**.

| Name  | Required  | Type | Description |
|---|---|--|--|
| `isActive`  | yes | [Boolean](#boolean) | A flag to indicate whether or not an account / participant / settlement model is active. |

### CurrencyEnum

For details, see section [7.3.9 Currency](../fspiop-api/documents/API-Definition_v1.1.md#739-currency) in the Mojaloop FSPIOP API Definition.

> ### PartyIdentifier

Reviewer: This isn't referenced elsewhere in this doc.

| `accounts`  |  yes | [Accounts](#accounts) | The list of ledger accounts configured for the participant. |

### ParticipantFunds

Data model for the complex type **ParticipantFunds**.

| Name  | Required  | Type | Description |
|---|---|--|--|
| `transferId`  | yes  | [UUID](#uuid) | Transfer identifier. |
| `externalReference`  |  yes | [String](#string) | Reference to any external data, such as an identifier from the settlement bank.  |
| `action`  |  yes | [Enum](#enum) | The action performed on the funds. Possible values are: `recordFundsIn` and `recordFundsOutPrepareReserve`. |
| `reason`  |  yes | [String](#string) | The reason for the FundsIn or FundsOut action. |
| `amount`  | yes  | [Money](#money) | The FundsIn or FundsOut amount. |
| `extensionList`  | no  | [ExtensionList](#extensionlist) | Additional details. |

### ParticipantLimit

Data model for the complex type **ParticipantLimit**.

| Name  | Required  | Type | Description |
|---|---|--|--|
| `type`  |  yes | [String](#string) | The type of participant limit (for example, `NET_DEBIT_CAP`.)  |
| `value`  |  yes | [Number](#number) | The value of the limit that has been set for the participant.  |

> | `alarmPercentage`  |  yes | [Number](#number) | An alarm notification is triggered when a pre-specified percentage of the limit is reached. Specifying an `alarmPercentage` is optional. If not specified, it will default to 10 percent. |

Reviewer: Is this a percentage expressed between 0-1? Or between 0->100?

e.g. does 0.5 mean 50%, or 0.5%?

Author: I believe it is between 0-100 but I'm not sure. @mdebarros, please can you help me out here? I will add this info to the doc once I have the answer.
